feat: add more scenarios to o3de manipulation benchmark

jmatejcz opened the pull request:
## Purpose
https://github.com/RobotecAI/rai/issues/433
- Make creation of new tasks faster
- Add more tasks for o3deTestBenchmark
- Add more simulation configs for o3deTestBenchmark

## Proposed Changes
- ManipulationTask for common logic
- Made tasks more generic and parametrized them so they can operate on different types of objects -> this will result in much more variants of a task available, for exampe: instead of move carrot to the left side -> move {object types} to the left side, where object types is passed as argument
- Refactored tasks -> MoveObjectToLeftTask, GroupObjectsTask
- New type of tasks -> BuildCubeTowerTask, PlaceObjectAtCoordTask, RotateObjectTask(not aplicable right now)
- New scenes configs, renamed them to specify what objects are present.
- Unit tests for helper functions and tasks to ensure proper score calculation 
- Packets of scenarios, that are subjectively grouped into 5 level of difficulty: trivial, easy, medium, hard, very_hard (can and probably will be adjusted). For now there are 10 trivial, 42 easy, 23 medium, 38 hard and 47 very hard scenarios.
- Resetting arm to base position
- Launching new binaries


## Issues 
- Not able to test manually if scores are calculated properly in different scenarios, as number of possible scenarios grows significantly. Depending on unit tests to ensure that scores are calculated properly. These tests must cover all cases.

- Not enough trivial scenarios (like move object to coords when few objects). Harder scenarios like creating structures from many objects can be created in various number of ways. It isn't the same for easy scenarios where the scenario should include 1, maybe 2 moves and scene setup that is not complicated. If you have ideas for other easy tasks, let me know in comments!

- For now ManipulatorMoveTo tool does not allow changing orientation of the gripper, which makes some tasks like RotateObjectTask, or scenes with rotated objects not usable.

- Corn entities are to small for gripper, always falling out, which result in them being avoided when defining tasks

- When gripper is holding an object it can place it into other object

## Testing

### Setup
1. Setup the repository
2. Install dependencies listed in: 
https://github.com/RobotecAI/rai/blob/main/docs/demos/manipulation.md

and:
```bash
poetry install --with openset
vcs import < demos.repos
rosdep install --from-paths src/examples/rai-manipulation-demo/ros2_ws/src --ignore-src -r -y
colcon build --symlink-install
source setup_shell.sh
```

3. Download GameLauncher binary: [humble](https://robotecai.sharepoint.com/:u:/g/EfBxGXkAOCVApgxD1QPW-nABov0bquxBgkIOJ_HGsdQhHA) or [jazzy](https://robotecai.sharepoint.com/:u:/g/ETTc3wvDM2xOjjySUjb-85YBHUW4g1MFi_69bbcR3iU3vw)

4. Populate `src/rai_bench/rai_bench/o3de_test_bench/configs/o3de_config.yaml` with:
```yaml
binary_path: /path/to/your/GameLauncher
level: RoboticManipulationBenchmark
robotic_stack_command: ros2 launch examples/manipulation-demo-no-binary.launch.py
required_simulation_ros2_interfaces:
  services:
    - /spawn_entity
    - /delete_entity
  topics:
    - /color_image5
    - /depth_image5
    - /color_camera_info5
  actions: []
required_robotic_ros2_interfaces:
  services:
    - /grounding_dino_classify
    - /grounded_sam_segment
    - /manipulator_move_to
  topics: []
  actions: []
robotic_stack_command: ros2 launch examples/manipulation-demo-no-binary.launch.py
required_simulation_ros2_interfaces:
  services:
    - /spawn_entity
    - /delete_entity
  topics:
    - /color_image5
    - /depth_image5
    - /color_camera_info5
  actions: []
required_robotic_ros2_interfaces:
  services:
    - /grounding_dino_classify
    - /grounded_sam_segment
    - /manipulator_move_to
  topics: []
  actions: []
```
### Run examples

Run the benchmark with example scenarios. I do not recommend running full packets as this will take some time. 
In `src/rai_bench/rai_bench/examples/o3de_test_benchmark.py` there are prepared packets of scenarios. 
You can swap running all_scenarios for for example 3 trivial scenarios -> t_scenarios[:3] here:
```python
    benchmark = Benchmark(
        simulation_bridge=o3de,
        scenarios=all_scenarios,
        logger=bench_logger,
        results_filename=results_filename,
    )
```
then:
```bash
python src/rai_bench/rai_bench/examples/o3de_test_benchmark.py
```

what to look for:
- check if all packets are running
- check how are they different, if you think the grading should be changed or some scenarios are not suited for the level, let me know in the comment
- logs and results can be found in  `src/rai_bench/rai_bench/experiments/`
- check if arm resets to base position properly after every scenario
### Tests
Run unit tests:
```bash
pytest tests/rai_bench/tasks/
```
- Check if they all pass
- Check the tests, if you can think of cases that are not covered, please let me know in comment.

jmatejcz committed:
feat: add ManipulationTask

jmatejcz committed:
feat: add helper funtions to find clusters in objects

jmatejcz committed:
feat: add GroupVegetablesTask

jmatejcz committed:
style: formatting change

jmatejcz committed:
chore: add licenses to new files

jmatejcz committed:
refactor: adjusted GroupVegetablesTask to rebased code

jmatejcz committed:
moved common logic to ManipulationTask

jmatejcz committed:
feat: added 5th scene

jmatejcz committed:
style: variables naming correction

jmatejcz committed:
feat: added build tower task

jmatejcz committed:
style: naming changes
type ignores added to imports

jmatejcz committed:
feat: add buildtower sub tasks

jmatejcz committed:
refactor: add Entity Generic to do calculations not only on Spawned Entities
Validating if there is any objects placed incorrectly, before creating scenario

jmatejcz committed:
style: adjust formatting

jmatejcz committed:
fix: config validation now correctly takes entites

jmatejcz committed:
feat: add new cube scenes

jmatejcz committed:
feat: parametrize MoveObjectsToLeftTask to take object types as argument

jmatejcz committed:
feat : parametrize GroupObjectsTask to take object types as argument

jmatejcz committed:
refactor: delete unused code from tasks

jmatejcz committed:
refactor: increased recusrsion limit for agent
add type hint for agent

jmatejcz committed:
feat: parametrize BuildTowetTask to take objects types as arguments

jmatejcz committed:
feat: add new scenes

jmatejcz committed:
fix: adjust validation in MoveObjectToLeftTask

jmatejcz committed:
test: basic test for Task helper methods

jmatejcz committed:
test: basic buildTowerTask tests

jmatejcz committed:
fix: adjust calculations in Task helper methods
fixed comparision of Entities

jmatejcz committed:
refactor:  score calculation in BuildCubeTowerTask

jmatejcz committed:
fix:  multiple towers test fix
common methods to conftest

jmatejcz committed:
refactor: distance parameters as args
docstrings for tasks
naming adjustments

jmatejcz committed:
test: basic GroupObjectsTask tests

jmatejcz committed:
fix: tasks now properly take all entities for calculations
adjusted calculations to cover other cases

jmatejcz committed:
fix: properly count initial objects

jmatejcz committed:
style: added docstrings to benchmark_model

jmatejcz committed:
refactor: moved distance args to tasks' constructors

jmatejcz committed:
feat: add place object at coords task with unit tests

jmatejcz committed:
test: add tests for place cubes task and move to left task

jmatejcz committed:
feat: add change orientation task with unit tests

jmatejcz committed:
feat: add new easy simulaiton configs

jmatejcz committed:
feat: group scenarios into trivial and easy packets

jmatejcz committed:
chore: deleted old scenes

jmatejcz committed:
style: added note about meters
formatting changes

jmatejcz committed:
feat: add more medium and hard simulation configs

jmatejcz committed:
feat: add scenarios packets of medium, hard, and very hard level

jmatejcz committed:
style: adjusted naming od rotateObjectTask file

jmatejcz committed:
feat: update README

jmatejcz committed:
style: adjust comments naming and logs

jmatejcz committed:
refactor: longer wait for stack readiness
prepered packets in benchmark example

boczekbartek commented:
@jmatejcz Thank you for the PR
I can see that "This branch is out-of-date with the base branch" - could you please "Update with rebase"?

jmatejcz commented:
> @jmatejcz Thank you for the PR I can see that "This branch is out-of-date with the base branch" - could you please "Update with rebase"?

done

boczekbartek commented:
@jmatejcz 
Thank you for this PR. I tried example commands, but didn't manage to run the benchmark. I think some nodes might not start correctly. 

Quick note: In step 2, before  running `colcon build` I did `vcs import < demos.repos; rosdep ...` to download `rai-manipulation-demo`.

I configured the path the the demo binary as described. When I run this command:

```bash
python src/rai_bench/rai_bench/examples/o3de_test_benchmark.py
```

I can see that simulator started, but some ros2 nodes seem to be missing:
```
2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required topics: []
2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required actions: []
2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO available actions: {'/move_action', '/panda_arm_controller/follow_joint_trajectory', '/panda_hand_controller/gripper_cmd', '/execute_trajectory'}
2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
```

Here is a full log: 
[log.log](https://github.com/user-attachments/files/19328655/log.log)

I think one of the reasons is that I get this error. Did you also encounter it?
```
$ ros2 launch examples/manipulation-demo-no-binary.launch.py
...
[grounded_sam-2] [ERROR] [1742334299.934854130] [grounded_sam]: Could not load model
```

Also, in the logs of the benchmark I can see a lot (402 in total) of logs like:
```
Could not create Scenario from task: Manipulate objects, so that ........
```
Are they expected? If they are expected - should user know all of them?

boczekbartek commented on src/rai_bench/rai_bench/examples/o3de_test_benchmark.py:
```diff
@@ -21,21 +21,32 @@
+    GetObjectPositionsTool,
+    MoveToPointTool,
+)
+from rai.tools.ros2.topics import (  # type: ignore
```
Please remove `# type: ignore` from imports. It looks like your IDE might not parse rai packages correctly. Probably python virtualenv is not connected properly. But there is no reason to ignore here, these imports are fine

jmatejcz replied:
i added it as strict pylance type checking throws -> 'Stub file not found for ...' . What level of type checking do you recommend?

maciejmajek replied:
I suggest using strict but taking it's warnings/errors with a grain of salt

boczekbartek commented on src/rai_bench/rai_bench/o3de_test_bench/tasks/build_tower_task.py:
```diff
@@ -0,0 +1,129 @@
+        TypeError
+            If any of the provided object types is not allowed.
+        """
+        # TODO (jm) what if allowable_displament is greater then the size of object?
```
How about this TODO?

jmatejcz replied:
I' m not sure how to treat and how to calculate cases when displacement is that large that it allows objects to be not stacked properly. This issue emerges, because Task does not know what is the size of the object, only coordinates of it's center. I had an idea to check if next object's 'z' coord is greater, but how much? i would need the size of the object to estimate that. If we assume any greater 'z' is valid, what about cases like: 
![buildtowertask1](https://github.com/user-attachments/assets/0c532638-23d6-4d8f-8e3a-aecacf30fdd1)
where the cube on top has greater z than the right cube so it can  be treated as valid tower on top of the right cube, because allowable displacement allows it.
or :
![buildtowertask2](https://github.com/user-attachments/assets/e5568884-0ce1-42e3-a4cf-2231cd68437c)
where the middle cube has greater 'z' but it doesn't form a proper tower.

So i postponed this issue, as maybe in future we will have access to objects sizes via SimulationBridge?

jmatejcz replied:
and allowable displacement is required as cubes won't be perfectly stacked

boczekbartek replied:
Thanks for the explanation! But in case 2 I think the center would be shifted on x/y axis and entities will no be grouped ([code](https://github.com/RobotecAI/rai/blob/577feaebd7055e9c7c15fda20d54faa8f73a571b/src/rai_bench/rai_bench/o3de_test_bench/tasks/build_tower_task.py#L112)), right?

Also, for O3DE Manipulation benchmark we can easily check the size of the available 3d models of cubes and set sensible [allowable_displacement](https://github.com/RobotecAI/rai/blob/577feaebd7055e9c7c15fda20d54faa8f73a571b/src/rai_bench/rai_bench/o3de_test_bench/tasks/build_tower_task.py#L34) in this task. It's manual checking in the O3DE Editor, but we can have a config for this task, that is well tuned to current o3de setup, even without support for object sizes by the `SimulationBridge`.

jmatejcz replied:
> Thanks for the explanation! But in case 2 I think the center would be shifted on x/y axis and entities will no be grouped ([code](https://github.com/RobotecAI/rai/blob/577feaebd7055e9c7c15fda20d54faa8f73a571b/src/rai_bench/rai_bench/o3de_test_bench/tasks/build_tower_task.py#L112)), right?

It depends on the allowable displacement, if it allows such big shift, they will be assigned to one group, like in the image, the center is in the range of allowable displacement.

jmatejcz replied:
> Also, for O3DE Manipulation benchmark we can easily check the size of the available 3d models of cubes and set sensible [allowable_displacement](https://github.com/RobotecAI/rai/blob/577feaebd7055e9c7c15fda20d54faa8f73a571b/src/rai_bench/rai_bench/o3de_test_bench/tasks/build_tower_task.py#L34) in this task. It's manual checking in the O3DE Editor, but we can have a config for this task, that is well tuned to current o3de setup, even without support for object sizes by the `SimulationBridge`.

Yes we can do it like this, but it would require providing this value separately for every object type and would limit the possibility of scaling the difficulty, as smaller allowable_displacement requires more precise stacking.

Though we could enforce upper limit like this, which makes sense to me

jmatejcz replied:
ideal solution would be getting object  size directly from simulation and scaling this value accordingly. Not sure if this is possible

jmatejcz replied:
as we discussed in person, i added maximal allowable displacement to object type
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/60a757744b2c0f42aa05d37a63157cc6bdfc6e52

boczekbartek commented on src/rai_bench/rai_bench/o3de_test_bench/tasks/grab_carrot_task.py:
Is this file removed on purpose?

jmatejcz replied:
yes, this task was refactored into MoveObjectToLeftTask

boczekbartek reviewed (commented):
@jmatejcz added some high-level comments/questions to the code

jmatejcz commented:
> @jmatejcz Thank you for this PR. I tried example commands, but didn't manage to run the benchmark. I think some nodes might not start correctly.
> 
> Quick note: In step 2, before running `colcon build` I did `vcs import < demos.repos; rosdep ...` to download `rai-manipulation-demo`.
> 
> I configured the path the the demo binary as described. When I run this command:
> 
> ```shell
> python src/rai_bench/rai_bench/examples/o3de_test_benchmark.py
> ```
> 
> I can see that simulator started, but some ros2 nodes seem to be missing:
> 
> ```
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required topics: []
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required actions: []
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO available actions: {'/move_action', '/panda_arm_controller/follow_joint_trajectory', '/panda_hand_controller/gripper_cmd', '/execute_trajectory'}
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
> ```
> 
> Here is a full log: [log.log](https://github.com/user-attachments/files/19328655/log.log)
> 
> I think one of the reasons is that I get this error. Did you also encounter it?
> 
> ```
> $ ros2 launch examples/manipulation-demo-no-binary.launch.py
> ...
> [grounded_sam-2] [ERROR] [1742334299.934854130] [grounded_sam]: Could not load model
> ```
> 
> Also, in the logs of the benchmark I can see a lot (402 in total) of logs like:
> 
> ```
> Could not create Scenario from task: Manipulate objects, so that ........
> ```
> 
> Are they expected? If they are expected - should user know all of them?

yes it seems like you can't properly load grounded dino and grounded sam models, which results in:
`WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']`

try removing any existing weights:
```bash
rm -rf build/ install/ log/
```
then once again:
```bash
colcon build --symlink-install
source setup_shell.sh
python src/rai_bench/rai_bench/examples/o3de_test_benchmark.py
```

jmatejcz commented:
> @jmatejcz Thank you for this PR. I tried example commands, but didn't manage to run the benchmark. I think some nodes might not start correctly.
> 
> Quick note: In step 2, before running `colcon build` I did `vcs import < demos.repos; rosdep ...` to download `rai-manipulation-demo`.
> 
> I configured the path the the demo binary as described. When I run this command:
> 
> ```shell
> python src/rai_bench/rai_bench/examples/o3de_test_benchmark.py
> ```
> 
> I can see that simulator started, but some ros2 nodes seem to be missing:
> 
> ```
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required topics: []
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO required actions: []
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] INFO available actions: {'/move_action', '/panda_arm_controller/follow_joint_trajectory', '/panda_hand_controller/gripper_cmd', '/execute_trajectory'}
> 2025-03-18 22:41:00 robo-pc-005 Agent logger[234244] WARNING Waiting for missing services ['/grounding_dino_classify', '/grounded_sam_segment'] out of required services: ['/grounding_dino_classify', '/grounded_sam_segment', '/manipulator_move_to']
> ```
> 
> Here is a full log: [log.log](https://github.com/user-attachments/files/19328655/log.log)
> 
> I think one of the reasons is that I get this error. Did you also encounter it?
> 
> ```
> $ ros2 launch examples/manipulation-demo-no-binary.launch.py
> ...
> [grounded_sam-2] [ERROR] [1742334299.934854130] [grounded_sam]: Could not load model
> ```
> 
> Also, in the logs of the benchmark I can see a lot (402 in total) of logs like:
> 
> ```
> Could not create Scenario from task: Manipulate objects, so that ........
> ```
> 
> Are they expected? If they are expected - should user know all of them?

to the second part of the question, the logs about `Could not create Scenario from task:` are expected but should be in debug level, i will adjust that. Thank you for the note.

boczekbartek commented:
@jmatejcz 
> `rm -rf build/ install/ log/`

I tested with a fresh install. Probalby gdino weights got corrupted for some reason... but it worked now. Thank you!

> to the second part of the question, the logs about Could not create Scenario from task: are expected but should be in debug level, I will adjust that. Thank you for the note.

Could you tell me a bit more about what this log message means?

jmatejcz commented:
> @jmatejcz
> 
> > `rm -rf build/ install/ log/`
> 
> I tested with a fresh install. Probalby gdino weights got corrupted for some reason... but it worked now. Thank you!
> 
> > to the second part of the question, the logs about Could not create Scenario from task: are expected but should be in debug level, I will adjust that. Thank you for the note.
> 
> Could you tell me a bit more about what this log message means?

Every task validates if given simulation config is suitable by checking if required objects are present and if any of them is placed incorrectly:
https://github.com/RobotecAI/rai/blob/jm/feat/o3de-bench-more-tasks/src/rai_bench/rai_bench/o3de_test_bench/tasks/manipulation_task.py#L81-L104
If not this means scenario wont be created out of these 2.

This is especially useful when scenarios are created automatically like stated in README. This means  you can pass a list of tasks, list of simulation configs and get every possible combination of task x sim_config as scenarios.

boczekbartek commented on src/rai_bench/README.md:
````diff
@@ -52,3 +48,32 @@ scenarios = Benchmark.create_scenarios(
+)
+```
+
+which are grouped by their subjective difficulty. For now there are 10 trivial, 42 easy, 23 medium, 38 hard and 47 very hard scenarios.
````
How do you measure the difficulty?

jmatejcz replied:
subjectively, you can check docstrings and code in `src/rai_bench/rai_bench/o3de_test_bench/scenarios.py` on how i assigned scenarios. They can be easily changed and extended if needed

boczekbartek replied:
I see, thanks for the pointer! Could you also clarify in the readme that dosctrings contain more explanation about the decision of grouping into difficulty categories?

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/0501d6ce045fc8d4fb4ac1dd42a56f765b4f3786

boczekbartek commented on src/rai_bench/rai_bench/examples/o3de_test_benchmark.py:
```diff
@@ -136,43 +147,37 @@
     benchmark = Benchmark(
         simulation_bridge=o3de,
-        scenarios=scenarios,
+        scenarios=t_scenarios,
```
I would include all scenarios by default

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/96bd95d307442d3778812f10236a7c84f52174e4

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -58,7 +60,14 @@ def __init__(
+        -------
+        bool
+            True if the configuration is suitable, False otherwise.
+        """
```
this docstring seems off, previous one was better

jmatejcz replied:
yes, thank you, i overlooked that

boczekbartek replied:
please share the commit link here when it's fixed

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/0501d6ce045fc8d4fb4ac1dd42a56f765b4f3786

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -268,22 +542,23 @@ def run_next(self, agent) -> None:
+            try:
+                result = scenario.task.calculate_result(self.simulation_bridge)
+                total_time = te - ts
+                self._logger.info(  # type: ignore
```
I think "type: ignore" is not useful here

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/577feaebd7055e9c7c15fda20d54faa8f73a571b

jmatejcz committed:
style: removed type ignores from imports and loggers
added logger to scenario
adjusted log about not creating scenario to debug level

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -134,10 +200,166 @@ def count_adjacent(self, positions: List[Pose], threshold_distance: float) -> in
+        Dict[str, List[EntityT]]
+            A dictionary with keys as prefab names and values as lists of entities of that type.
+        """
+        entities_by_type: Dict[str, List[EntityT]] = {}
```
I think it's more clear than `.setdefault`
```suggestion
        entities_by_type: Dict[str, List[EntityT]] = defaultdict(list)
```

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/57d8569506156d3f62ca3af698aadc5114be9708

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -134,10 +200,166 @@ def count_adjacent(self, positions: List[Pose], threshold_distance: float) -> in
+        return clusters
+
+    def group_entities_along_z_axis(
+        # TODO (jm) figure out how to group by other coords and orientation, without reapeting code
```
It's better to use you github username in todos. Also TODO suggests that this should be done before merging. 
How about changing this TODO to NOTE if you don't plan to implement it now?

jmatejcz replied:
yes, you are right, I removed some todos and changed other to notes, applied here:
https://github.com/RobotecAI/rai/pull/452/commits/57d8569506156d3f62ca3af698aadc5114be9708

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -134,10 +200,166 @@ def count_adjacent(self, positions: List[Pose], threshold_distance: float) -> in
+
+        return clusters
+
+    def group_entities_along_z_axis(
```
I would like to ask if I understand this method correctly.

If I have 2 towers of boxes in the scene, I should get 2 groups of entities here, right?

jmatejcz replied:
yes, if docstring is not clear, please let me know how would you change it

boczekbartek replied:
I think if you could include the 2 tower example in the docstring that would be perfect ⭐

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/6f15158002b10a9a83535e9c6cb7f92243254085

boczekbartek replied:
thanks!

boczekbartek commented on tests/rai_bench/tasks/test_task.py:
```diff
@@ -0,0 +1,116 @@
+
+    groups: List[List[Entity]] = task.group_entities_along_z_axis(entities, margin=0.1)
+    assert len(groups) == 1
+    assert groups[0] == [e1, e2, e3, e4, e5]
```
```suggestion
    assert groups[0] == entities
```

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/57d8569506156d3f62ca3af698aadc5114be9708

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -134,10 +200,166 @@ def count_adjacent(self, positions: List[Pose], threshold_distance: float) -> in
+            entities_by_type.setdefault(entity.prefab_name, []).append(entity)
+        return entities_by_type
+
+    def check_neighbourhood_types(
```
This method is only used in tests. Do you think it's useful to keep it in a `Task`? If not, I would move it to tests as a separate utility there

jmatejcz replied:
i think it might be useful, for now i just don't use it in any task. Maybe in future tasks it will be used

boczekbartek replied:
So for example if we would have a task like: "surround a cube with carrots" we could check if carrots are near the cube?

jmatejcz replied:
Yes, exactly. It might seem redundant with find_clusters method, but the example you provided shows it can be useful in some cases

boczekbartek reviewed (changes requested):
@jmatejcz I've added some comments to the code, but didn't mange to check all of it yet. I'll continue

jmatejcz committed:
style: adjusted doctrings and README

jmatejcz committed:
refactor: applied suggested changes to code and notes

jmatejcz committed:
refactor: run all scenarios by default

jmatejcz committed:
style: add example to group_entities_along_z_axis
removed redundant NOTE

boczekbartek commented on src/rai_bench/README.md:
````diff
@@ -52,3 +48,33 @@ scenarios = Benchmark.create_scenarios(
+```
+
+which are grouped by their subjective difficulty. For now there are 10 trivial, 42 easy, 23 medium, 38 hard and 47 very hard scenarios.
+Check docstrings and code in src/rai_bench/rai_bench/o3de_test_bench/scenarios.py if you want to know how scenarios are assigned to difficulty level.
````
Please make `src/rai_bench/rai_bench/o3de_test_bench/scenarios.py` a link.
It would be good to replace all file paths in this readme with links to files in the repo, so it's easier to navigate. 
Here is an example of how to do it: [link](https://github.com/RobotecAI/rai/blob/6f15158002b10a9a83535e9c6cb7f92243254085/README.md?plain=1#L143)

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/c9133f50d26ec8af47caad4b8fdfb2c1ab8276aa

boczekbartek commented on src/rai_bench/rai_bench/benchmark_model.py:
```diff
@@ -208,12 +475,11 @@ def create_scenarios(
-                    print(
-                        f"Could not create Scenario from task: {task.get_prompt()} and simulation_config: {sim_conf}, {e}"
-                    )
+                except ValueError:
```
It would be more clear to validate the config before scenario is created, rather than catch an exception here.
It's just that the config is not suitable for a task, rather than error handling

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/f36cf53bc08815f45bfd7cda873ea3a3a8b6b839

jmatejcz committed:
style: add links in README

boczekbartek commented on src/rai_bench/rai_bench/examples/o3de_test_benchmark.py:
```diff
@@ -136,43 +147,39 @@
+        translation=Translation(x=0.1, y=0.5, z=0.4),
+        rotation=Rotation(x=1.0, y=0.0, z=0.0, w=0.0),
+    )
+    for i in range(len(t_scenarios)):
```
```suggestion
    for i in range(len(all_scenarios)):
```

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/0165900976ce3328b4519eec494707113aee80dd

boczekbartek commented on src/rai_bench/rai_bench/o3de_test_bench/tasks/move_object_to_left_task.py:
```diff
@@ -0,0 +1,77 @@
+
+    def get_prompt(self) -> str:
+        obj_names = ", ".join(obj + "s" for obj in self.obj_types).replace("_", " ")
+        return f"Manipulate objects, so that all of the following objects are on the left side of the table (positive y): {obj_names}."
```
I am not sure if the prompt can be understood well.
I would replace "the following objects" with `{obj_names}` to make it more human.
Also I am not sure about `(positive y)` part.

jmatejcz replied:
Adjusted the prompt, keeping the y part as the interpretation of 'left side' it still depends on the camera position and i'm not sure how to resolve this. Added note to keep it in mind. 
Applied here:
https://github.com/RobotecAI/rai/pull/452/commits/5f9af31a975824cf818552ebec44b337f4e7aa3a

boczekbartek commented on src/rai_bench/rai_bench/examples/o3de_test_benchmark.py:
```diff
@@ -136,43 +147,39 @@
     )
-    for i, s in enumerate(scenarios):
+    # custom request to arm
+    base_arm_pose = Pose(
```
I think we don't ensure that  `base_arm_pose` has been reached after each task.
I noticed that sometimes arm is in a blocked state and moveit can't fix it. 
I would check if the base_arm_pose is achieved and restart the game launcher if there are problems
Otherwise benchmark task will fail, because moveit can't figure out the trajectory.

jmatejcz replied:
@knicked will create feature that allows to reset arm and all joints to base position, so waiting for that

jmatejcz replied:
https://github.com/RobotecAI/rai-manipulation-demo/pull/8

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/999d0e19603dac1ec2fc08b89a9840fe64703271

boczekbartek commented:
@jmatejcz I didn't run all the tasks myself, but I think next step after this PR is to select the final set of tasks. Besides tasks and llm/rai capabilities that they check we should take into account execution time. 
Could you share here execution times from your experiments?

boczekbartek commented:
@jmatejcz Response to [this message](https://github.com/RobotecAI/rai/pull/452#issuecomment-2735813744).
Yes, debug log should be enough for this information.

jmatejcz committed:
fix: loop though all scenarios

jmatejcz committed:
refactor: changed MoveObjectsToLeftTask prompt to be more understandable
added note about the positive y part

jmatejcz committed:
refactor: moved validating config before creating scenario

jmatejcz commented:
> @jmatejcz I didn't run all the tasks myself, but I think next step after this PR is to select the final set of tasks. Besides tasks and llm/rai capabilities that they check we should take into account execution time. Could you share here execution times from your experiments?

i didn't run all myself too as this will take very long time.
From my experiments i can tell that easier scenarios require less time ~30second-1minute, harder scenarios can take even up to 3 minutes. Sometimes agent won't do task at all which will take much less time.

For example running 29 trivial and easy tasks took only ~15minutes ~30 second per task:
[results.csv](https://github.com/user-attachments/files/19342029/results.csv)

and 30 very hard tasks took ~28 minutes , ~1minute per task, but it will certainly grow as in majority of scenarios agent didn't perform whole task  .
[results.csv](https://github.com/user-attachments/files/19342634/results.csv)

boczekbartek commented:
@jmatejcz Thanks! Thats very useful info!
> From my experiments i can tell that easier scenarios require less time ~30second-1minute, harder scenarios can take even 
up to 3 minutes. Sometimes agent won't do task at all which will take much less time.

Also, I think hard task timeout should be added, I don't see it in the code right now.

jmatejcz committed:
refactor: added maximum allowable displacement to buildTowerTask

boczekbartek commented on src/rai_bench/README.md:
```diff
@@ -1,36 +1,32 @@
-For more information about these classes go to -> `src/rai_bench/rai_bench/benchmark_model.py`
+For more information about these classes go to -> [benchmark_model](./rai_bench/benchmark_model.py)
 
 ### O3DE TEST BENCHMARK
```
Why CAPS LOCK?

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/e94542434a50a413e24be47dad3cf17cd3174ca1

boczekbartek commented on src/rai_bench/README.md:
```diff
@@ -1,36 +1,32 @@
-O3DE Test Benchmark (`src/rai_bench/rai_bench/o3de_test_bench/`), contains 2 Tasks(`tasks/`) - GrabCarrotTask and PlaceCubesTask (these tasks implement calculating scores) and 4 scene_configs(`configs/`) for O3DE robotic arm simulation.
-
-Both tasks calculate score, taking into consideration 4 values:
+The O3DE Test Benchmark [o3de_test_benchmark_module](./rai_bench/o3de_test_bench/) provides tasks and scene configurations for robotic arm manipulation task. The tasks use a common `ManipulationTask` logic and can be parameterized, which allows for many task variants. The current tasks include:
```
I think O3DE binary release for benchmarks should be mentioned in the readme.

jmatejcz replied:
it is not available publicly for now

jmatejcz replied:
applied here:
https://github.com/RobotecAI/rai/pull/452/commits/3e3abee7e21350621c14db020713cfc9726dff58

jmatejcz commented:
> @jmatejcz Thanks! Thats very useful info!
> 
> > From my experiments i can tell that easier scenarios require less time ~30second-1minute, harder scenarios can take even
> > up to 3 minutes. Sometimes agent won't do task at all which will take much less time.
> 
> Also, I think hard task timeout should be added, I don't see it in the code right now.

I can't think of an easy and fast solution right now, it would require probably terminating whole simulation.
I will add it as a separate issue

https://github.com/RobotecAI/rai/issues/462

boczekbartek commented:
@jmatejcz Please change links from the PR description to public links: 
> Download GameLauncher binary: [humble](https://robotecai.sharepoint.com/:u:/g/EfBxGXkAOCVApgxD1QPW-nABov0bquxBgkIOJ_HGsdQhHA) or [jazzy](https://robotecai.sharepoint.com/:u:/g/ETTc3wvDM2xOjjySUjb-85YBHUW4g1MFi_69bbcR3iU3vw)

jmatejcz committed:
style: README capslock title removed

jmatejcz committed:
test: add tests for not allowed displacement and types to buildTowerTask

jmatejcz commented:
adjusted lunching for new binaries:
https://github.com/RobotecAI/rai/pull/452/commits/bbc6bc2b5e0e43f2aff1e845fda86cef9ec1e632

@MagdalenaKotynia  please check this change, as this is part of rai_sim package

Changed o3de_config.yaml, so don't forget to update it. I edited description of PR to include new field in config

jmatejcz committed:
refactor: adjust launching binaries to match new binaries

MagdalenaKotynia commented:
@jmatejcz The change https://github.com/RobotecAI/rai/commit/bbc6bc2b5e0e43f2aff1e845fda86cef9ec1e632 looks good. When this PR is merged, I will add the example content of` o3de_config.yaml` with links to appropriate binaries to rai_sim README.

jmatejcz committed:
feat: reseting arm to base position after scenario

jmatejcz commented:
added resetting arm to base position which is available here: https://github.com/RobotecAI/rai-manipulation-demo/pull/8
applied here: https://github.com/RobotecAI/rai/pull/452/commits/999d0e19603dac1ec2fc08b89a9840fe64703271

Please check once again if everything works with new binaries and resetting arm.

jmatejcz committed:
feat: include binaries links in README

maciejmajek commented:
@boczekbartek @MagdalenaKotynia are you familiar with the state of this PR? How can we move it forward?

MagdalenaKotynia commented:
@maciejmajek I am not familiar with the state of this PR, I haven't reviewed this PR, except for https://github.com/RobotecAI/rai/commit/bbc6bc2b5e0e43f2aff1e845fda86cef9ec1e632 (I left [a comment](https://github.com/RobotecAI/rai/pull/452#issuecomment-2742809979), that it looks good).

boczekbartek commented:
@maciejmajek 
> are you familiar with the state of this PR? How can we move it forward?

It needs final tests after https://github.com/RobotecAI/rai/pull/452/commits/999d0e19603dac1ec2fc08b89a9840fe64703271, both on ubuntu 22.04 and 24.04. 
I am just about to test it, I'll update you soon.

boczekbartek committed:
fix: arm reset and cleanup in o3de manipulation benchmark   (#481)

boczekbartek committed:
feat: add example config

boczekbartek committed:
docs: improve benchmark readme

boczekbartek reviewed (approved):
Approved.

- tested on ubuntu 22.04 & ubuntu 24.04
- added minor changes to this PR:
  - added config from the PR description to the repo
  - adjusted `rai_bench/README`
  - refactored `/reset_manipulator` service call to use `ROS2Connector` api

boczekbartek committed:
chore: pre-commit
